Default values for const generic parameters MAX_DIRS, MAX_FILES, don't work as expected.

Commit https://github.com/rust-embedded-community/embedded-sdmmc-rs/commit/566f2023653dace195a43b05a6429f47e6be8a50 added two const generic parameters `MAX_DIRS` and `MAX_FILES` to `Controller`.

Those parameters have default values equal to the values used before, and the new docs say: "By default the `Controller` will initialize with a maximum number of `4` open directories and files. This can be customized by specifying the `MAX_DIR` and `MAX_FILES` generic consts of the `Controller`"

However, the default values don't work as expected. The following code causes a compile failure:
```
let mut cont = Controller::new(block, DummyTimesource::default());
```
```
error[E0282]: type annotations needed for `embedded_sdmmc::Controller<BlockSpi<'_, rp_pico::rp2040_hal::Spi<rp_pico::rp2040_hal::spi::Enabled, SPI0, 8>, rp_pico::rp2040_hal::gpio::Pin<Gpio5, rp_pico::rp2040_hal::gpio::Output<rp_pico::rp2040_hal::gpio::PushPull>>>, DummyTimesource, MAX_DIRS, MAX_FILES>`
   --> boards/rp-pico/examples/pico_spi_sd_card.rs:276:9
    |
276 |     let mut cont = Controller::new(block, DummyTimesource::default());
    |         ^^^^^^^^
    |
help: consider giving `cont` an explicit type, where the the value of const parameter `MAX_DIRS` is specified
    |
276 |     let mut cont: embedded_sdmmc::Controller<BlockSpi<'_, rp_pico::rp2040_hal::Spi<rp_pico::rp2040_hal::spi::Enabled, SPI0, 8>, rp_pico::rp2040_hal::gpio::Pin<Gpio5, rp_pico::rp2040_hal::gpio::Output<rp_pico::rp2040_hal::gpio::PushPull>>>, DummyTimesource, MAX_DIRS, MAX_FILES> = Controller::new(block, DummyTimesource::default());
    |                 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

For more information about this error, try `rustc --explain E0282`.
```
(This was noticed in `rp-hal`, see [this ticket](https://github.com/rp-rs/rp-hal/pull/426) for context.)

There is a blog post which nicely explains why this happens: https://gankra.github.io/blah/defaults-affect-inference/#default-type-parameters

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -9,11 +9,6 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 [Unreleased]: https://github.com/rust-embedded-community/embedded-sdmmc-rs/compare/v0.4.0...develop
 
-### Changes
-- Add `MAX_DIRS` and `MAX_FILES` generics to `Controller` to allow an arbitrary numbers of concurrent open directories and files.
-- Add new constructor method `Controller::new_with_limits(block_device: D, timesource: T) -> Controller<D, T, MAX_DIRS, MAX_FILES>`
-  to create a `Controller` with custom limits.
-
 ## [Version 0.4.0](https://github.com/rust-embedded-community/embedded-sdmmc-rs/releases/tag/v0.4.0)
 
 ### Changes
@@ -27,6 +22,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Fix `env_logger` pulling in the `std` feature in `log` in library builds.
 - Raise the minimum supported Rust version to 1.56.0.
 - Code tidy-ups and more documentation.
+- Add `MAX_DIRS` and `MAX_FILES` generics to `Controller` to allow an arbitrary numbers of concurrent open directories and files.
+- Add new constructor method `Controller::new_with_limits(block_device: D, timesource: T) -> Controller<D, T, MAX_DIRS, MAX_FILES>`
+  to create a `Controller` with custom limits.
 
 ## [Version 0.3.0](https://github.com/rust-embedded-community/embedded-sdmmc-rs/releases/tag/v0.3.0)
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -11,6 +11,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ### Changes
 - Add `MAX_DIRS` and `MAX_FILES` generics to `Controller` to allow an arbitrary numbers of concurrent open directories and files.
+- Add new constructor method `Controller::new_custom_max(block_device: D, timesource: T) -> Controller<D, T, MAX_DIRS, MAX_FILES>`
+  to create a `Controller` with custom limits.
 
 ## [Version 0.4.0](https://github.com/rust-embedded-community/embedded-sdmmc-rs/releases/tag/v0.4.0)
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -11,7 +11,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ### Changes
 - Add `MAX_DIRS` and `MAX_FILES` generics to `Controller` to allow an arbitrary numbers of concurrent open directories and files.
-- Add new constructor method `Controller::new_custom_max(block_device: D, timesource: T) -> Controller<D, T, MAX_DIRS, MAX_FILES>`
+- Add new constructor method `Controller::new_with_limits(block_device: D, timesource: T) -> Controller<D, T, MAX_DIRS, MAX_FILES>`
   to create a `Controller` with custom limits.
 
 ## [Version 0.4.0](https://github.com/rust-embedded-community/embedded-sdmmc-rs/releases/tag/v0.4.0)

--- a/README.md
+++ b/README.md
@@ -42,7 +42,7 @@ let mut cont: Controller<
     DummyTimeSource,
     6,
     12,
-> = Controller::new(block, time_source);
+> = Controller::new_custom_max(block, time_source);
 ```
 
 ## Supported features

--- a/README.md
+++ b/README.md
@@ -42,7 +42,7 @@ let mut cont: Controller<
     DummyTimeSource,
     6,
     12,
-> = Controller::new_custom_max(block, time_source);
+> = Controller::new_with_limits(block, time_source);
 ```
 
 ## Supported features

--- a/examples/create_test.rs
+++ b/examples/create_test.rs
@@ -118,7 +118,7 @@ fn main() {
         .map_err(Error::DeviceError)
         .unwrap();
     println!("lbd: {:?}", lbd);
-    let mut controller: Controller<LinuxBlockDevice, Clock, 4, 4> = Controller::new(lbd, Clock);
+    let mut controller = Controller::new(lbd, Clock);
     for volume_idx in 0..=3 {
         let volume = controller.get_volume(VolumeIdx(volume_idx));
         println!("volume {}: {:#?}", volume_idx, volume);

--- a/examples/delete_test.rs
+++ b/examples/delete_test.rs
@@ -118,7 +118,7 @@ fn main() {
         .map_err(Error::DeviceError)
         .unwrap();
     println!("lbd: {:?}", lbd);
-    let mut controller: Controller<LinuxBlockDevice, Clock, 4, 4> = Controller::new(lbd, Clock);
+    let mut controller = Controller::new(lbd, Clock);
     for volume_idx in 0..=3 {
         let volume = controller.get_volume(VolumeIdx(volume_idx));
         println!("volume {}: {:#?}", volume_idx, volume);

--- a/examples/test_mount.rs
+++ b/examples/test_mount.rs
@@ -121,7 +121,7 @@ fn main() {
         .map_err(Error::DeviceError)
         .unwrap();
     println!("lbd: {:?}", lbd);
-    let mut controller: Controller<LinuxBlockDevice, Clock, 4, 4> = Controller::new(lbd, Clock);
+    let mut controller = Controller::new(lbd, Clock);
     for i in 0..=3 {
         let volume = controller.get_volume(VolumeIdx(i));
         println!("volume {}: {:#?}", i, volume);

--- a/examples/write_test.rs
+++ b/examples/write_test.rs
@@ -118,7 +118,7 @@ fn main() {
         .map_err(Error::DeviceError)
         .unwrap();
     println!("lbd: {:?}", lbd);
-    let mut controller: Controller<LinuxBlockDevice, Clock, 4, 4> = Controller::new(lbd, Clock);
+    let mut controller = Controller::new(lbd, Clock);
     for volume_idx in 0..=3 {
         let volume = controller.get_volume(VolumeIdx(volume_idx));
         println!("volume {}: {:#?}", volume_idx, volume);

--- a/src/controller.rs
+++ b/src/controller.rs
@@ -40,10 +40,10 @@ where
     /// files.
     ///
     /// This creates a Controller with default values
-    /// MAX_DIRS = 4, MAX_FILES = 4. Call `Controller::new_custom_max(block_device, timesource)`
+    /// MAX_DIRS = 4, MAX_FILES = 4. Call `Controller::new_with_limits(block_device, timesource)`
     /// if you need different limits.
     pub fn new(block_device: D, timesource: T) -> Controller<D, T, 4, 4> {
-        Self::new_custom_max(block_device, timesource)
+        Self::new_with_limits(block_device, timesource)
     }
 }
 
@@ -56,7 +56,10 @@ where
     /// Create a new Disk Controller using a generic `BlockDevice`. From this
     /// controller we can open volumes (partitions) and with those we can open
     /// files.
-    pub fn new_custom_max(block_device: D, timesource: T) -> Controller<D, T, MAX_DIRS, MAX_FILES> {
+    pub fn new_with_limits(
+        block_device: D,
+        timesource: T,
+    ) -> Controller<D, T, MAX_DIRS, MAX_FILES> {
         debug!("Creating new embedded-sdmmc::Controller");
         Controller {
             block_device,

--- a/src/controller.rs
+++ b/src/controller.rs
@@ -29,6 +29,24 @@ where
     open_files: [(VolumeIdx, Cluster); MAX_FILES],
 }
 
+impl<D, T> Controller<D, T, 4, 4>
+where
+    D: BlockDevice,
+    T: TimeSource,
+    <D as BlockDevice>::Error: core::fmt::Debug,
+{
+    /// Create a new Disk Controller using a generic `BlockDevice`. From this
+    /// controller we can open volumes (partitions) and with those we can open
+    /// files.
+    ///
+    /// This creates a Controller with default values
+    /// MAX_DIRS = 4, MAX_FILES = 4. Call `Controller::new_custom_max(block_device, timesource)`
+    /// if you need different limits.
+    pub fn new(block_device: D, timesource: T) -> Controller<D, T, 4, 4> {
+        Self::new_custom_max(block_device, timesource)
+    }
+}
+
 impl<D, T, const MAX_DIRS: usize, const MAX_FILES: usize> Controller<D, T, MAX_DIRS, MAX_FILES>
 where
     D: BlockDevice,
@@ -38,7 +56,7 @@ where
     /// Create a new Disk Controller using a generic `BlockDevice`. From this
     /// controller we can open volumes (partitions) and with those we can open
     /// files.
-    pub fn new(block_device: D, timesource: T) -> Controller<D, T, MAX_DIRS, MAX_FILES> {
+    pub fn new_custom_max(block_device: D, timesource: T) -> Controller<D, T, MAX_DIRS, MAX_FILES> {
         debug!("Creating new embedded-sdmmc::Controller");
         Controller {
             block_device,

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -460,8 +460,9 @@ mod tests {
 
     #[test]
     fn partition0() {
-        let mut c: Controller<DummyBlockDevice, Clock, 4, 4> =
-            Controller::new(DummyBlockDevice, Clock);
+        let mut c: Controller<DummyBlockDevice, Clock, 2, 2> =
+            Controller::new_custom_max(DummyBlockDevice, Clock);
+
         let v = c.get_volume(VolumeIdx(0)).unwrap();
         assert_eq!(
             v,

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -461,7 +461,7 @@ mod tests {
     #[test]
     fn partition0() {
         let mut c: Controller<DummyBlockDevice, Clock, 2, 2> =
-            Controller::new_custom_max(DummyBlockDevice, Clock);
+            Controller::new_with_limits(DummyBlockDevice, Clock);
 
         let v = c.get_volume(VolumeIdx(0)).unwrap();
         assert_eq!(

